From 55ddfd50856f9cd9bb25ad8282714e46ba9c063d Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:37 -0400
Subject: [ethernet] mlx4: mark napi id for gro_skb

Message-id: <633e127cb799659e38a762d633c61ddac07f36f2.1410425016.git.avadai@redhat.com>
Patchwork-id: 93709
O-Subject: [Patch RHEL7.1 V1 161/197] mlx4: mark napi id for gro_skb
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Jason Wang <jasowang@redhat.com>

Napi id was not marked for gro_skb, this will lead rx busy loop won't
work correctly since they stack never try to call low latency receive
method because of a zero socket napi id. Fix this by marking napi id
for gro_skb.

The transaction rate of 1 byte netperf tcp_rr gets about 50% increased
(from 20531.68 to 30610.88).

Cc: Amir Vadai <amirv@mellanox.com>
Signed-off-by: Jason Wang <jasowang@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 32b333fe99069d090a12cc106fd5ae1002fe642f)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_rx.c b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
index edf6189..3b62b0b 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_rx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
@@ -782,6 +782,7 @@ int mlx4_en_process_rx_cq(struct net_device *dev, struct mlx4_en_cq *cq, int bud
             PKT_HASH_TYPE_L3);
 
      skb_record_rx_queue(gro_skb, cq->ring);
+     skb_mark_napi_id(gro_skb, &cq->napi);
 
      if (ring->hwtstamp_rx_filter == HWTSTAMP_FILTER_ALL) {
       timestamp = mlx4_en_get_cqe_ts(cqe);
-- 
1.7.1